modules/tutorial-advanced-minimalist-fulcro/pages/index.adoc

```diff
@@ -0,0 +1,338 @@
+.. The component can use the `:will-enter` component option function to access the route parameters (which are always strings) and to translate them into the actual ident via `(dr/route-immediate <ident>)` or `route-deferred`
+3. Initialize everything and then route to the target you want displayed via `dr/change-route`
+
+Let's explore a minimalist example of routing. This is the UI that we are creating:
+
+image::bare-essentials-routing-ui-mockup.png["A mock up of a UI with a router (via wiredjs.com)"]
+
+image::../images/bare-essentials-routing-ui-mockup.png["FIXME: remove"]
+
++++<!--FIXME: Remove the relative path img above-->+++
+
+We show either all the people or just a single person. This is the code:
+
+.Bare essentials of Fulcro routing (1st half - router & targets)
+====
+```clojure
+(defsc AllPeople [_ {:keys [all-people]}]
+  {:ident (fn [] [:component/id ::AllPeople])
+   :query [{:all-people [:person/id :person/name]}]  ; <1>
+   :initial-state {}                                 ; <1>
+   :route-segment ["all"]}                           ; <2>
+   (dom/div
+     (dom/h3 "All People")
+     (dom/ul
+       (mapv (fn [{:person/keys [id name]}] (dom/li {:key id} name))
+         all-people))))
+
+(defsc Person [_ {:person/keys [id name biography]}]
+  {:ident :person/id
+   :query [:person/id :person/name :person/biography]  ; <1>
+   :initial-state {}                                   ; <1>
+   :route-segment ["person" :person-id]                ; <3>
+   :will-enter (fn [app route-params]                  ; <4>
+                 (dr/route-immediate
+                   [:person/id
+                    (js/parseInt (:person-id route-params))]))}
+  (dom/p (str "Person #" id ": ") (dom/strong name) " - " biography))
+
+(dr/defrouter MyRouter [_ _] {:router-targets [AllPeople Person]}) ; <5>
+```
+====
+<1> A router target _must always_ have a query (use the "give me everything" query `['*]` if it doesn't need any) and a non-nil initial state (so at least an empty `{}`)
+<2> A target's `:route-segment` assigns it a routing path (relative to an ancestor target, if any)
+<3> The `:route-segment` may also contain placeholders to capture parameters from the route (here `:person-id`),
+    which can then be accessed in `:will-enter` under that keyword
+<4> A target may use the `:will-enter` component option function to access route params and to tell Fulcro which ident
+    to route to (here via `dr/route-immediate`)
+<5> `defrouter` creates the router component. It needs at least the list of its targets in `:router-targets`
+
+WARNING: will-enter can be called multiple times as part of the route resolution algorithm and MUST NOT side-effect 
```

Comment:
Hukka: but the examples do IO, and the book says "Finally, since the will-enter on any give router can do I/O it is never a good idea for the library to automatically trigger routes because that could result in unexpected I/O and side-effects for off-screen components."

JH: I guess I need to clarify. You must not do any I/O inside will-enter itself. If you want any, it must be inside the completion function, as in the example. Is this clearer?

Reply:
The book also does explicitly forbid the side-effects too in another part, so perhaps this is more of a problem in the main Fulcro doc than in this tutorial. Still, it is a bit confusing when reading all of this for the first time. Might work better to explicitly say that the side-effects must be in a lambda passed to the deferred.